### ISSUE TITLE
${workspaceFolder} variable is not expanded before passing include paths to cppcheck

### ISSUE THREAD

**Issue by bootress** [jbenden/vscode-c-cpp-flylint#38]
I've reduced it to a minimal example for the purposes of this report, I have a project in a Windows environment:

```
test_proj/
  src/
    test_source.c
  include/
    test_header.h
````

I've added the `include` directory as an extra include path to the default ones given by the C/Cpp extension:
```
c_cpp_properties.json
...
"includePath": [
  "${workspaceFolder}",
  "${workspaceFolder}/include",
  "C:/Program Files (x86)/Microsoft Visual Studio 14.0/VC/include",
  "C:/Program Files (x86)/Microsoft Visual Studio 14.0/VC/atlmfc/include",
  "C:/Program Files (x86)/Windows Kits/10/Include/10.0.10240.0/um",
  "C:/Program Files (x86)/Windows Kits/10/Include/10.0.10240.0/ucrt",
  "C:/Program Files (x86)/Windows Kits/10/Include/10.0.10240.0/shared"
],
...
```

When I save `test_source.c`, the debug output from the extension is:
```
Performing lint scan of c:\Users\skearney\test_proj\src\test_source.c...
Error expanding include path '${workspaceFolder}': undefined
executing:  C:\Program Files\Cppcheck\cppcheck.EXE --inline-suppr --enable=warning,style,performance,portability,information -I /include -I C:/Program Files (x86)/Microsoft Visual Studio 14.0/VC/include -I C:/Program Files (x86)/Microsoft Visual Studio 14.0/VC/atlmfc/include -I C:/Program Files (x86)/Windows Kits/10/Include/10.0.10240.0/um -I C:/Program Files (x86)/Windows Kits/10/Include/10.0.10240.0/ucrt -I C:/Program Files (x86)/Windows Kits/10/Include/10.0.10240.0/shared --std=c11 --std=c++11 -D_DEBUG -DUNICODE -D_UNICODE --language=c++ --platform=native --template={file}  {line}  {severity} {id}: {message} c:\Users\skearney\test_proj\src\test_source.c
[ '(information) Couldn\'t find path given by -I \'\\include\\\'',
  'Checking c:\\Users\\skearney\\test_proj\\src\\test_source.c ...',
  'Checking c:\\Users\\skearney\\test_proj\\src\\test_source.c: _DEBUG=1;UNICODE=1;_UNICODE=1...',
  '' ]
[ '    information missingInclude: Cppcheck cannot find all the include files (use --check-config for details)',
  '' ]
An exception occurred in CppCheck while parsing output for file c:\Users\skearney\test_proj\src\test_source.c stacktrace: Error: Line could not be parsed: (information) Couldn't find path given by -I '\include\'
    at CppCheck.parseLine (C:\Users\skearney\.vscode\extensions\jbenden.c-cpp-flylint-0.6.0\server\linters\cppcheck.js:68:19)
    at lines.forEach.line (C:\Users\skearney\.vscode\extensions\jbenden.c-cpp-flylint-0.6.0\server\linters\linter.js:208:31)
    at Array.forEach (native)
    at CppCheck.parseLines (C:\Users\skearney\.vscode\extensions\jbenden.c-cpp-flylint-0.6.0\server\linters\linter.js:207:15)
    at CppCheck.lint (C:\Users\skearney\.vscode\extensions\jbenden.c-cpp-flylint-0.6.0\server\linters\linter.js:195:25)
    at lintersCopy.forEach.linter (C:\Users\skearney\.vscode\extensions\jbenden.c-cpp-flylint-0.6.0\server\server.js:104:39)
    at Array.forEach (native)
    at validateTextDocument (C:\Users\skearney\.vscode\extensions\jbenden.c-cpp-flylint-0.6.0\server\server.js:101:17)
    at documents.onDidSave (C:\Users\skearney\.vscode\extensions\jbenden.c-cpp-flylint-0.6.0\server\server.js:63:9)
    at CallbackList.invoke (C:\Users\skearney\.vscode\extensions\jbenden.c-cpp-flylint-0.6.0\server\node_modules\vscode-jsonrpc\lib\events.js:71:39)
Completed lint scans...
```

It's cool that the extension is grabbing my include paths from `c_cpp_properties.json` and passing them to cppcheck, but is it possible for `${workspaceFolder}` to be expanded?

**Comment by jbenden** [comment]
It is possible.

This variable is a relatively new addition to VS Code, introduced with multiple folders inside a single workspace.

In order for the extension to become aware of the multiple folders, some reworking of the extension will be needed, which may break backwards compatibly with older versions of VS Code. Note, however, that I am not 100% sure on it breaking backwards compatibility, until researching the VS Code language server protocol.

-Joe

**Comment by bootress** [comment]
Interesting. I have only been using VS Code for a short while, but I've only used single-folder workspaces. The ${workspaceFolder} variable was in the c_cpp_properties.json template that the editor created, so I assumed it was meant to be used to locate relative paths inside the folder you have opened with VS Code.

Thanks for looking into this.

**Comment by jbenden** [comment]
Problem should be resolved in the latest version, just published.

Best regards,
-Joe
